From 634a0aff7cc1be91962c771504fcc1fd29da89b4 Mon Sep 17 00:00:00 2001
From: Kevin Funk <kevin.funk@kdab.com>
Date: Tue, 3 Jan 2017 22:37:30 +0100
Subject: [PATCH] CMake: More meaningful warning if CMP0025 not set

When using non-Apple Clang on macOS with CMP0025 not set to new, CMake 3
is not able to deduce the available compiler features and errors out
with e.g.:

 CMake Error in src/CMakeLists.txt:
   No known features for CXX compiler

   "Clang"

   version 3.9.0.

This is easily fixed by setting the CMP0025 policy to NEW.
Detect this early, and print an appropriate warning to the user.

Task-number: QTBUG-57886
Change-Id: I792a43a3612b1fe6e0781310e58a9839c29d3b44
---
 src/corelib/Qt5CoreConfigExtras.cmake.in | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/qtbase/src/corelib/Qt5CoreConfigExtras.cmake.in b/qtbase/src/corelib/Qt5CoreConfigExtras.cmake.in
index 545b9a3..e1f2de8 100644
--- a/qtbase/src/corelib/Qt5CoreConfigExtras.cmake.in
+++ b/qtbase/src/corelib/Qt5CoreConfigExtras.cmake.in
@@ -114,6 +114,32 @@ set(Qt5_DISABLED_FEATURES
 
 set_property(TARGET Qt5::Core APPEND PROPERTY INTERFACE_COMPILE_DEFINITIONS $<$<NOT:$<CONFIG:Debug>>:QT_NO_DEBUG>)
 
+# print more meaningful error if we run into QTBUG-57886
+#if (APPLE AND POLICY CMP0025)
+#    # note: if CMP0025 available => there's at least CMake 3.0 available
+#    cmake_policy(GET CMP0025 _cmp0025_state)
+#    get_property(_cxx_known_features GLOBAL PROPERTY CMAKE_CXX_KNOWN_FEATURES)
+#    list(FIND _cxx_known_features \"cxx_decltype\" _index)
+#    if (NOT _cmp0025_state STREQUAL \"NEW\" AND _index EQUAL -1 AND CMAKE_CXX_COMPILER_ID STREQUAL \"Clang\")
+#        message(WARNING \"CMake can\'t detect compiler features properly if CMP0025 is not set to NEW and non-Apple \"
+#            \"Clang is used (see QTBUG-57886). Consider adding this to the beginning of your root CMakeLists.txt:\n\"
+#            \"  cmake_policy(SET CMP0025 NEW)\")
+#    endif()
+#endif()
+if (APPLE AND POLICY CMP0025)
+    # note: if CMP0025 available => there's at least CMake 3.0 available
+    cmake_policy(GET CMP0025 _cmp0025_state)
+    get_property(_cxx_known_features GLOBAL PROPERTY CMAKE_CXX_KNOWN_FEATURES)
+    list(FIND _cxx_known_features \"cxx_decltype\" _index)
+    if (CMAKE_CXX_COMPILER_ID STREQUAL \"Clang\")
+        if (_cmp0025_state STREQUAL \"OLD\" OR (NOT _cmp0025_state STREQUAL \"NEW\" AND CMAKE_MINIMUM_REQUIRED_VERSION VERSION_LESS 3.0))
+            message(WARNING \"CMake cannot detect compiler features properly if CMP0025 is not set to NEW and non-Apple \"
+                \"Clang is used (see QTBUG-57886). Consider adding this to the beginning of your root CMakeLists.txt:\n\"
+                \"  cmake_policy(SET CMP0025 NEW)\")
+        endif()
+    endif()
+endif()
+
 set_property(TARGET Qt5::Core PROPERTY INTERFACE_COMPILE_FEATURES cxx_decltype)
 
 !!IF qtConfig(reduce_exports)
